source/common/grpc/async_client_manager_impl.cc

```diff
@@ -84,8 +86,24 @@ GoogleAsyncClientFactoryImpl::GoogleAsyncClientFactoryImpl(
   // Check metadata for gRPC API compliance. Uppercase characters are lowered in the HeaderParser.
   // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
   for (const auto& header : config.initial_metadata()) {
-    if (!validateGrpcHeaderChars(header.key()) || !validateGrpcHeaderChars(header.value())) {
-      throw EnvoyException("Illegal characters in gRPC initial metadata.");
+    // Validate key
+    if (!validateGrpcHeaderChars(header.key())) {
+      throw EnvoyException(fmt::format("Illegal characters in gRPC initial metadata header key: {}.",header.key()));
+    }
+
+    // Validate value
+    // Binary base64 encoded
+    if (::absl::EndsWith(header.key(), "-bin")) {
```

Comment:
If I'm understanding this code right, it looks like this is requiring that the metadata be already base64 encoded.  I don't think that's necessary for GOOGLE_GRPC, because the gRPC library will automatically do base64 encoding and decoding for header values if the header name ends with "-bin".

Reply:
What happens if it is already base64 encoded?

Reply:
If the gRPC library automatically does encoding/decoding - what check should be done in this case? or is any value allowed?

Reply:
We don't check to see if the value is already base64-encoded; we just unconditionally base64-encode it.  So it would wind up being double-base64-encoded on the wire.

I think this will actually work, but it does add unnecessary overhead.

Reply:
Then we can probably allow any value for `-bin`. This check was added in https://github.com/envoyproxy/envoy/pull/14129 (CC @asraa) because fuzzers were crashing the gRPC library with inputs that looked like https://github.com/envoyproxy/envoy/pull/14129/files#diff-2a20e9888f070730d0412bf9806728cfae9f9f9269747837f68287ff4fcdb5caR1.

Reply:
That seems right, any value probably goes for `-bin`. 
Otherwise you would have seen a crash in the grpc library in the test GoogleGrpcIllegalLengthBase64 if you didn't catch with the exception below...

On the PR as a whole, thanks for correcting this check!